### DISCUSSION TITLE
What's the difference between grep and live_grep?

### DISCUSSION THREAD

**Discussion by niamleeson** [ibhagwan/fzf-lua#1139] [General]
And why isn't there a grep_project equivalent, like live_grep_project?

**Comment by ibhagwan** [comment]
Read the below explore the generated commands with `debug=true` and you’ll understand
https://github.com/ibhagwan/fzf-lua/wiki#whats-the-difference-between-grep-and-live_grep

**Reply by niamleeson** [reply]
Thank you! What does it mean when it says fzf is used as UI selector only?

**Reply by ibhagwan** [reply]
It means that no fuzzy matching is done, each keypress generates a new rg command with what you typed being used as the regex input to `rg`.

**Comment by MartenBE** [comment]
@ibhagwan I understand the difference between `grep` and `livegrep`, but have difficulty to see how this translates to `grep_curbuf` and `lgrep_curbuf`? When using debug, these seem equal?

**Reply by ibhagwan** [reply]
> When using debug, these seem equal?

Not at all, using grep_curbuf is equal to grep the rg command never changes and only fuzzy match is performed, when you switch to lgrep_curbuf you’re switching to live_grep and every key would generate a new command with your input as thr regex, if the debug is the same you’re doing something wrong or not understanding what you’re seeing.

**Reply by MartenBE** [reply]
Anyone who is still confused about it, I tried it out with an example and it clarified the FAQ entry a lot:

`dummy.txt`

```txt
Lorem ipsum dolor sit amet, consectetur adipiscing elit.
    Sed sed diam auctor, feugiat justo id, congue nisi.
    Quisque laoreet nisl vitae leo vehicula auctor.
    Etiam eget augue sed magna congue feugiat.
    In sed arcu tempus, euismod elit eget, vehicula sem.

    Donec mattis arcu eget quam congue imperdiet.
    Suspendisse ultricies justo at urna laoreet pretium.
    Aliquam at augue dapibus, vestibulum risus sit amet, tincidunt quam.

    Nullam quis mauris dictum, pharetra turpis id, fringilla lacus.
    Proin sodales nibh nec finibus iaculis.
    Etiam eget nunc ut elit elementum auctor nec eu ipsum.
    Quisque imperdiet quam interdum, consequat ante dictum, porta tortor.
    Duis ultrices libero at lorem eleifend, id dapibus arcu iaculis.

    In efficitur leo commodo, tincidunt est vitae, efficitur lectus.
    Suspendisse eu odio vitae risus sollicitudin tincidunt vehicula ac enim.
    Morbi convallis purus sed mollis pellentesque.
    Integer mattis elit id turpis tempus rutrum.

    Quisque finibus mauris ut nisi scelerisque, a semper massa rutrum.
```

- `grep_curbuf` is the same as `blines`

    ![Screenshot_20250121_204231](https://github.com/user-attachments/assets/351670e1-dd63-429f-bab1-b8e564913f48)
    
    ![Screenshot_20250121_205237](https://github.com/user-attachments/assets/a00388ff-811b-4854-ac4f-997422d482d7)

- `lgrep_curbuf`

    ![Screenshot_20250121_204204](https://github.com/user-attachments/assets/9949582d-bc44-4645-8674-7b5a40e2a9ce)

The difference?

- `grep_curbuf` dumps all the bufferlines in fzf, allowing you to fuzzy find. It is essentially `cat dummy.txt | fzf` if the current buffer contains `dummy.txt`.
- `lgrep_curbuf` uses only `rg` and thus does no initial fuzzy finding.

You can see this if you look for line 3 of `dummy.txt` in the results. `grep_curbuf` has line 3 because there is an `i` somewhere followed by an `n` somewhere in the line. Not necessarily `in`, but fuzzy. This line does not correspond with the regexp `in` however and thus is line 3 not a result in `lgrep_curbuf`. It would be if the regexp were something like `i.*n`.

In hindsight, the [FAQ](https://github.com/ibhagwan/fzf-lua/wiki#whats-the-difference-between-grep-and-live_grep) and the hints in the UI are very clear about this @ibhagwan, I just needed an example to wrap my head around it and make it "click" mentally.

**Reply by ibhagwan** [reply]
> In hindsight, the [FAQ](https://github.com/ibhagwan/fzf-lua/wiki#whats-the-difference-between-grep-and-live_grep) and the hints in the UI are very clear about this @ibhagwan, I just needed an example to wrap my head around it and make it "click" mentally.

@MartenBE the wiki is open edit feel free to add your 2 cents there.

> grep_curbuf is the same as blines

Not fully accurate, grep requires buffer saved to the disk, blines uses the buffer (dirty or otherwise).

**Reply by MartenBE** [reply]
Ok, I have updated the wiki.

**Reply by ibhagwan** [reply]
> Ok, I have updated the wiki.

Ty @MartenBE, looks great, one thing I’m not sure if clear enough is that the live grep supports regexes (as it’s sent directly to rg), for example try `in.*?\s`

**Reply by MartenBE** [reply]
Ok, I clarified that a bit more

**Reply by ibhagwan** [reply]
> Ok, I clarified that a bit more

Ty @MartenBE!
